File: engine/Sync.cpp
#include "sys.h"
#include "Sync.hpp"

#include "ActionQueue.hpp"
#include "Logger.hpp"
#include "RemoteAction.hpp"
#include "Transport.hpp"
#include "Sys.hpp"


// The existing adventure port doesn't do a lot of dynamic memory allocation, but instead 
// allocates everything up front.  We'll stick to that paradigm as much as possible.

const int Sync::MAX_MESSAGE_SIZE = 256;

Sync::Sync(int inNumPlayers, int inThisPlayer, Transport* inTransport) :
    numPlayers(inNumPlayers),
    thisPlayer(inThisPlayer),
    gameWon(NULL),
    transport(inTransport),
    frameNum(0)
{
    playersLastMove = new PlayerMoveAction*[numPlayers];
    for(int ctr=0; ctr<numPlayers; ++ctr) {
        playersLastMove[ctr] = NULL;
    }
    msgsRcvdFromPlayer = new int[numPlayers];
    resetMessagesReceived();
}

Sync::~Sync() {
    delete[] playersLastMove;
}

void Sync::StartFrame() {
	++frameNum;
}

int Sync::getFrameNumber() {
    return frameNum;
}

int Sync::getMessagesReceived(int player) {
    return msgsRcvdFromPlayer[player];
}

void Sync::resetMessagesReceived() {
    for(int ctr=0; ctr<numPlayers; ++ctr) {
        msgsRcvdFromPlayer[ctr] = 0;
    }
}

void Sync::handled(RemoteAction* action) {
    // Record we got a message from the sender
    if (action != NULL) {
        int sender = action->sender;
        if ((sender >= 0) && (sender < numPlayers)) {
            ++msgsRcvdFromPlayer[sender];
        }
    }
}

void Sync::BroadcastAction(RemoteAction* action) {
    if (action != NULL) {
        action->setSender(thisPlayer);
        action->serialize(sendBuffer, MAX_MESSAGE_SIZE);
        if (transport != NULL) {
            transport->sendPacket(sendBuffer);
        }
        
        Logger::log() << "Sent \"" << sendBuffer << "\" on frame #" << frameNum << Logger::EOM;
        
        delete action;
    }
}

void Sync::RejectMessage(const char* message, const char* errorMsg) {
    Logger::logError() << "Cannot process message - " << errorMsg << ": " << message << Logger::EOM;
}

void Sync::handleBatMoveMessage(const char* message) {
    BatMoveAction* nextAction = new BatMoveAction();
    nextAction->deserialize(receiveBuffer);
    batMoves.enQ(nextAction);
    handled(nextAction);
}

void Sync::handleBatPickupMessage(const char* message) {
    BatPickupAction* nextAction = new BatPickupAction();
    nextAction->deserialize(receiveBuffer);
    batMoves.enQ(nextAction);
    handled(nextAction);
}

void Sync::handlePlayerMoveMessage(const char* message) {
    PlayerMoveAction* nextAction = new PlayerMoveAction();
    nextAction->deserialize(receiveBuffer);
    int messageSender = nextAction->sender;
    if (playersLastMove[messageSender] != NULL) {
        delete playersLastMove[messageSender];
    }
    playersLastMove[messageSender] = nextAction;
    handled(nextAction);
}

void Sync::handleDragonMoveMessage(const char* message) {
    DragonMoveAction* nextAction = new DragonMoveAction();
    nextAction->deserialize(receiveBuffer);
    dragonMoves.enQ(nextAction);
    handled(nextAction);
}

void Sync::handleDragonStateMessage(const char* message) {
    DragonStateAction* nextAction = new DragonStateAction();
    nextAction->deserialize(receiveBuffer);
    dragonMoves.enQ(nextAction);
    handled(nextAction);
}

void Sync::handlePlayerPickupMessage(const char* message) {
    PlayerPickupAction* nextAction = new PlayerPickupAction();
    nextAction->deserialize(receiveBuffer);
    playerPickups.enQ(nextAction);
    handled(nextAction);
}

void Sync::handlePlayerResetMessage(const char* message) {
    PlayerResetAction* nextAction = new PlayerResetAction();
    nextAction->deserialize(receiveBuffer);
    playerResets.enQ(nextAction);
    handled(nextAction);
}

void Sync::handlePortcullisStateMessage(const char* message) {
    PortcullisStateAction* nextAction = new PortcullisStateAction();
    nextAction->deserialize(receiveBuffer);
    gateStateChanges.enQ(nextAction);
    handled(nextAction);
}

void Sync::handlePlayerWinMessage(const char* message) {
    // Don't know how we'd get this, but we ignore any win message after we receive the first one.
    if (gameWon == NULL) {
        PlayerWinAction* nextAction = new PlayerWinAction();
        nextAction->deserialize(receiveBuffer);
        gameWon = nextAction;
        handled(nextAction);
    }
}

void Sync::handlePingMessage(const char* message) {
    PingAction* nextAction = new PingAction();
    nextAction->deserialize(receiveBuffer);
    // Don't need to do anything with the ping except mark that it was handled.
    handled(nextAction);
}

void Sync::handleMazeSetupObjectMessage(const char* message) {
    ObjectMoveAction* nextAction = new ObjectMoveAction();
    nextAction->deserialize(receiveBuffer);
    mazeSetupActions.enQ(nextAction);
    handled(nextAction);
}

int Sync::pullNextPacket(char* buffer, int bufferSize) {
    int numChars = (transport == NULL ? 0 : transport->getPacket(buffer, bufferSize));
    if (numChars > 0) {
        Logger::log() << "Received \"" << buffer << "\" on frame #" << frameNum << Logger::EOM;
    }
    return numChars;
}


void Sync::PullLatestMessages() {
    const int TYPE_LENGTH = 2;
    int numChars = pullNextPacket(receiveBuffer, MAX_MESSAGE_SIZE);
    while(numChars >= TYPE_LENGTH+1) {
        char type[TYPE_LENGTH+1];
        // First two characters are the type of message
        for(int ctr=0; ctr<TYPE_LENGTH; ++ctr) {
            type[ctr] = receiveBuffer[ctr];
        }
        switch (receiveBuffer[0]) {
            case 'P':
                switch (receiveBuffer[1]) {
                    case 'M': {
                        handlePlayerMoveMessage(receiveBuffer);
                        break;
                    }
                    case 'P': {
                        handlePlayerPickupMessage(receiveBuffer);
                        break;
                    }
                    case 'R': {
                        handlePlayerResetMessage(receiveBuffer);
                        break;
                    }
                    case 'W': {
                        handlePlayerWinMessage(receiveBuffer);
                        break;
                    }
                    default:
                        printf("Message with unknown message type P%c: %s\n", receiveBuffer[1], receiveBuffer);
                }
                break;
            case 'B':
                switch (receiveBuffer[1]) {
                    case 'M': {
                        handleBatMoveMessage(receiveBuffer);
                        break;
                    }
                    case 'P': {
                        handleBatPickupMessage(receiveBuffer);
                        break;
                    }
                    default:
                        printf("Message with unknown message type B%c: %s\n", receiveBuffer[1], receiveBuffer);
                }
                break;
            case 'D':
                switch (receiveBuffer[1]) {
                    case 'M': {
                        handleDragonMoveMessage(receiveBuffer);
                        break;
                    }
                    case 'S': {
                        handleDragonStateMessage(receiveBuffer);
                        break;
                    }
                    default:
                        printf("Message with unknown message type D%c: %s\n", receiveBuffer[1], receiveBuffer);
                }
                break;
            case 'G':
                switch (receiveBuffer[1]) {
                    case 'S': {
                        handlePortcullisStateMessage(receiveBuffer);
                        break;
                    }
                    default:
                        printf("Message with unknown message type G%c: %s\n", receiveBuffer[1], receiveBuffer);
                }
                break;
            case 'M':
                switch (receiveBuffer[1]) {
                    case 'O': {
                        handleMazeSetupObjectMessage(receiveBuffer);
                        break;
                    }
                    default:
                        printf("Message with unknown message type M%c: %s\n", receiveBuffer[1], receiveBuffer);
                }
                break;
            case 'X':
                switch (receiveBuffer[1]) {
                    case 'X': {
                        handlePingMessage(receiveBuffer);
                        break;
                    }
                    default:
                        printf("Message with unknown message type X%c: %s\n", receiveBuffer[1], receiveBuffer);
                }
                break;
            default:
                printf("Message with unknown message type %c*: %s\n", receiveBuffer[0], receiveBuffer);
        }
        
        numChars = pullNextPacket(receiveBuffer, MAX_MESSAGE_SIZE);
    }
}

PlayerMoveAction* Sync::GetLatestBallSync(int player) {
    PlayerMoveAction* rtn = playersLastMove[player];
    playersLastMove[player] = NULL;
    return rtn;
}

PlayerResetAction* Sync::GetNextResetAction() {
    RemoteAction* next = NULL;
    if (!playerResets.isEmpty()) {
        next = playerResets.deQ();
    }
    return (PlayerResetAction*)next;
}

RemoteAction* Sync::GetNextDragonAction() {
    RemoteAction* next = NULL;
    if (!dragonMoves.isEmpty()) {
        next = dragonMoves.deQ();
    }
    return next;
}

PortcullisStateAction* Sync::GetNextPortcullisAction() {
    RemoteAction* next = NULL;
    if (!gateStateChanges.isEmpty()) {
        next = gateStateChanges.deQ();
    }
    return (PortcullisStateAction*)next;
}

RemoteAction* Sync::GetNextBatAction() {
    RemoteAction* next = NULL;
    if (!batMoves.isEmpty()) {
        next = batMoves.deQ();
    }
    return next;
}

PlayerPickupAction* Sync::GetNextPickupAction() {
    PlayerPickupAction* next = NULL;
    if (!playerPickups.isEmpty()) {
        next = (PlayerPickupAction*)playerPickups.deQ();
    }
    return next;

}

PlayerWinAction* Sync::GetGameWon() {
    PlayerWinAction* next = gameWon;
    gameWon = NULL;
    return next;
}

ObjectMoveAction* Sync::GetNextSetupAction() {
    ObjectMoveAction* next = NULL;
    if (!mazeSetupActions.isEmpty()) {
        next = (ObjectMoveAction*)mazeSetupActions.deQ();
    }
    return next;
    
}

File: H2HAdventure/Assets/Scripts/GameEngine/Sync.cs
using System;
using System.Collections.Generic;
namespace GameEngine
{
    public class Sync
    {
        private List<RemoteAction> batActions = new List<RemoteAction>();
        private List<RemoteAction> dragonActions = new List<RemoteAction>();
        private List<PlayerPickupAction> playerPickups = new List<PlayerPickupAction>();
        private List<PlayerResetAction> playerResets = new List<PlayerResetAction>();
        private List<PortcullisStateAction> gateStateChanges = new List<PortcullisStateAction>();
        private List<ObjectMoveAction> mazeSetupActions = new List<ObjectMoveAction>();
        private PlayerMoveAction[] playersLastMove;
        private PlayerWinAction gameWon;

        private const int CHECK_PERIOD = 15000; // Fifteen seconds

        private Transport transport;

        private int thisPlayer;

        private int numPlayers;

        private int[] msgsRcvdFromPlayer;

        public Sync(int inNumPlayers, int inThisPlayer, Transport inTransport)
        {
            numPlayers = inNumPlayers;
            thisPlayer = inThisPlayer;
            gameWon = null;
            transport = inTransport;
            playersLastMove = new PlayerMoveAction[numPlayers];
            for (int ctr = 0; ctr < numPlayers; ++ctr)
            {
                playersLastMove[ctr] = null;
            }
            msgsRcvdFromPlayer = new int[numPlayers];
            resetMessagesReceived();
        }

        /**
         * This pulls messages off the socket until there are none waiting.
         * This does not process them, but demuxes them and puts them where they can be grabbed
         * when it is time to process that type of message.
         */
        public void PullLatestMessages()
        {
            RemoteAction nextAction = (transport != null ? transport.get() : null);

            while (nextAction != null)
            {
                bool hitError = false;
                switch (nextAction.typeCode)
                {
                    case ActionType.PLAYER_MOVE:
                        int messageSender = nextAction.sender;
                        playersLastMove[messageSender] = (PlayerMoveAction)nextAction;
                        break;
                    case ActionType.PLAYER_PICKUP:
                        playerPickups.Add((PlayerPickupAction)nextAction);
                        break;
                    case ActionType.PLAYER_RESET:
                        playerResets.Add((PlayerResetAction)nextAction);
                        break;
                    case ActionType.PLAYER_WIN:
                        // Don't know how we'd get this, but we ignore any win message after we receive the first one.
                        if (gameWon == null)
                        {
                            gameWon = (PlayerWinAction)nextAction;
                        }
                        break;
                    case ActionType.DRAGON_MOVE:
                    case ActionType.DRAGON_STATE:
                        dragonActions.Add(nextAction);
                        break;
                    case ActionType.BAT_MOVE:
                    case ActionType.BAT_PICKUP:
                        batActions.Add(nextAction);
                        break;
                    case ActionType.PORTCULLIS_STATE:
                        gateStateChanges.Add((PortcullisStateAction)nextAction);
                        break;
                    case ActionType.OBJECT_MOVE:
                        mazeSetupActions.Add((ObjectMoveAction)nextAction);
                        break;
                    case ActionType.PING:
                        break;
                    default:
                        // Hit unknown type
                        hitError = true;
                        break;
                }
                if (!hitError)
                {
                    handled(nextAction);
                }

                nextAction = transport.get();
            }
        }


        /**
         *  Get the latest changes to another player.  Returns the last known state of a
         * player including their position and velocity.  Caller must delete this object.
         * If no changes have been received since the last call, this will return
         * null.
         */
        public PlayerMoveAction GetLatestBallSync(int player)
        {
            PlayerMoveAction rtn = playersLastMove[player];
            playersLastMove[player] = null;
            return rtn;
        }

        /**
         * Get the next player pickup or player drop action.
         * Caller must delete this action.
         * If no actions have been received, this will return null.
         */
        public PlayerPickupAction GetNextPickupAction()
        {
            PlayerPickupAction next = null;
            if (playerPickups.Count > 0)
            {
                next = playerPickups[0];
                playerPickups.RemoveAt(0);
            }
            return next;

        }

        /**
         * Get the next player reset action.
         * Caller must delete this action.
         * If no actions have been received, this will return null.
         */
        public PlayerResetAction GetNextResetAction()
        {
            PlayerResetAction next = null;
            if (playerResets.Count > 0)
            {
                next = playerResets[0];
                playerResets.RemoveAt(0);
            }
            return next;
        }

        /**
         * If another player has won, this will return that action.
         * Otherwise will return null.
         * Caller must delete this action.
         */
        public PlayerWinAction GetGameWon()
        {
            PlayerWinAction next = gameWon;
            gameWon = null;
            return next;
        }

        /**
         * Get the next dragon action.  Caller must delete this object.
         * Caller must delete this action.
         * If no actions have been received, this will return null.
         */
        public RemoteAction GetNextDragonAction()
        {
            RemoteAction next = null;
            if (dragonActions.Count > 0)
            {
                next = dragonActions[0];
                dragonActions.RemoveAt(0);
            }
            return next;
        }

        /**
         * Get the next portcullis action.  Caller must delete this object.
         * Caller must delete this action.
         * If no actions have been received, this will return null.
         */
        public PortcullisStateAction GetNextPortcullisAction()
        {
            PortcullisStateAction next = null;
            if (gateStateChanges.Count > 0)
            {
                next = gateStateChanges[0];
                gateStateChanges.RemoveAt(0);
            }
            return next;
        }

        /**
         * Get the next bat action.  Caller must delete this object.
         * Caller must delete this action.
         * If no actions have been received, this will return null.
         */
        public RemoteAction GetNextBatAction()
        {
            RemoteAction next = null;
            if (batActions.Count > 0)
            {
                next = batActions[0];
                batActions.RemoveAt(0);
            }
            return next;
        }

        /**
         * Get the next maze setup action.
         * Caller must delete this action.
         * If no actions have been received, this will return null.
         */
        public ObjectMoveAction GetNextSetupAction()
        {
            ObjectMoveAction next = null;
            if (mazeSetupActions.Count > 0)
            {
                next = mazeSetupActions[0];
                mazeSetupActions.RemoveAt(0);
            }
            return next;
        }

        /**
         * Broadcast an event to the other players
         * @param action an action to broadcast.  The Sync now owns this action and is responsible
         * for deleting it.
         */
        public void BroadcastAction(RemoteAction action)
        {
            if (action != null)
            {
                action.setSender(thisPlayer);
                if (transport != null)
                {
                    transport.send(action);
                }
            }

        }


        public int getMessagesReceived(int player)
        {
            return msgsRcvdFromPlayer[player];
        }

        public void resetMessagesReceived()
        {
            for (int ctr = 0; ctr < numPlayers; ++ctr)
            {
                msgsRcvdFromPlayer[ctr] = 0;
            }
        }

        private void handled(RemoteAction action)
        {
            // Record we got a message from the sender
            if (action != null)
            {
                int sender = action.sender;
                if ((sender >= 0) && (sender < numPlayers))
                {
                    ++msgsRcvdFromPlayer[sender];
                }
            }
        }

    }
}
